Verify method

Legogizmo opened the pull request:
Added a verify method ( #126 ).
And fixed the logic within `nwipe_runmethod()` fixing #17 #91

PartialVolume commented:
Will review shortly, however can you resubmit without the 2nd merge from the master commit appearing in the timeline. Makes for twice as many commits for stuff that's already been committed. Just makes the commit timeline more complicated than it needs to be. I guess you'll probably have to use rebase to do that on your existing local branch and resubmit a pull request.

For future commits, without having to rebase and in order to merge the master into your branch, is to create a second branch i.e VerifyMethod_2 based on the upstream master, i.e. your master that doesn't contain you patch.. yet. then merge VerifyMethod into VerifyMethod2 then do a pull request of that branch. i.e if your branch is VerifyMethod and you want to update with with a new upstream master then 

> git checkout master
git pull upstream master (so you local master is up to date)
git checkout -b VerifyMethod_2 (create a new branch based on the upstream master)
git checkout VerifyMethod_2 (switch to your new branch)
git merge VerifyMethod
Now do the pull request using the new VerifyMethod_2 branch and it won't show any master merge request.

I've not had to use the above method yet, although I may need to use it in the next couple of days. Looks pretty straight forward, however on your existing branch now you have merged the master you will either have to rebase or remove the master merge commit and then follow the above method. Any problems let me know and I'll see if I can help. Good discussion [here](https://stackoverflow.com/questions/16955980/git-merge-master-into-feature-branch/42666870)

Legogizmo commented:
@PartialVolume thanks for the tip, I'm still figuring out how this works, I'm using the github web interface . I'll figure out how to fix this, also I found a bug in my implementation with the nwipe_ops2 method verifying twice when verify last is selected. It will be fixed when I resubmit.

PartialVolume commented:
Not necessarily to be fixed for this pr but as a note to review for later.

1. When selecting method 'verify only' the blanking and verification are left on not switched off, is this correct ? You can manually switching blanking & verification off.
2. If you start the verify and and then see there are immediate verify errors you probably won't want to continue so you control-C out. Currently there is no indication in the log of the number of errors reported or an explicit statement saying the verify was aborted. This is a general problem with most methods and as I'm working in that area it's probably best if I deal with that issue.
3. On completion of a verify there are two places in the GUI that report the number of errors. There is a big discrepancy between both values, why is that ? I would have expected them to be the same.
![nwipe_MethodVerify_branch_error_discrepancy](https://user-images.githubusercontent.com/22084881/68777689-f9a53100-0629-11ea-8590-8af857986b26.png)
4. On completion of a full verify, the logs are shown below with confirmation that there are verification errors. However the number of errors doesn't seem to make sense. This test was conducted on a 80MB drive, that had previously had a full PRNG wipe with no blanking pass so I would have thought that the errors would have been of the order of ~19K (using 4096 block size) or 156K (using 512)
![nwipe_MethodVerify_log_on_full_verify](https://user-images.githubusercontent.com/22084881/68778582-4a695980-062b-11ea-8ed3-157e34fae4eb.png)

Legogizmo commented:
1) selecting 'verify only' doesn't effect the state of 'noblanking' or 'verify last' options. Instead I changed the logic in the `nwipe_runmethod()` this fixed the other errors resulting from conflicting options.
I added another `else if` condition after the nwipe_ops2 check to see if the method is 'verify only' that will run regardless of other options, I also changed the 'noblanking' check in that same area to also check if the method was 'zero'.

3 + 4) no clue why that is, but it is an already existing bug (guess I should submit that issue)

PartialVolume commented:
> edit: Looks like the web interface is less than ideal, switching to the command line.

Unless there is some reason you need to merge in the master there is often no need to do it. Just create your branch and ignore the fact the master has moved on with other commits. If there is a conflict when you do the pull request you can just leave it for me to sort out in the web interface.

I should really document the github flow I use, I think it's what other people do but could be wrong. I develop all my code using KDE's Kdevelop IDE. Basically these are the steps.

> Create a local branch when working on a bug
git checkout -b 337-fix-locking-scroll (make changes to code)
git status -uno
git commit -m "fixes issue #337"
git push origin 337-fix-locking-scroll (pushes your branch to the personal github fork on the web
git checkout master (change back to the master)
git pull upstream master (update your local copy from the master at https://github.com/martijnvanbrummelen/nwipe/)
git branch -d 337-fix-locking-scroll (removes branch, you need to be on a different branch)
git branch (to see what branch your on)
git branch * (to see all available branches)
git log (list commits)

One thing I always do when working on the command line to to make sure  you know which branch your on before typing a command, saves getting in a right mess, Im certainly no expert on github usage as it wasn't that long ago I knew nothing about git.

> git branch
>  Add_Serial_No_To_GUI_For_USB_Drives
>  certificate
>  exclude_drive_validation_check
>   **fix_segfault_on_CNTRL-C_at_end_wipe_take2**
>  last_pass_not_verified
>  master
>  memory_leak_c2
>  netinet_header_removal
>  nwipex
>  old_master

There's lots of other commands but basically those are all I tend to use, then occasionally I might have to use rebase to merge two or more commits into a single commit which is the preferred way before you submit a pr. My first ever pr on nwipe was a mess, historically recorded here for posterity, where I put about 20 fixes into a single branch, definitely not what you want to do ! :-)

Legogizmo commented:
That's good to know, I actually found a way to do it via the web interface (merge the original with the fork, go back a commit so the merge doesn't show, and branch from that).

So now I have a new branch that only has my one commit, should I close this one and submit that now?

PartialVolume commented:
Yes, go ahead and do that. Thanks.
